sylvia-derive/src/message.rs

```diff
@@ -14,12 +15,242 @@ use syn::parse::{Parse, Parser};
+                    let ty = field.ty;
+                    quote! { #name : #ty}
+                });
+
+                quote! {
+                    pub fn #function_name(mut self, #(#fields),*) -> Self {
+                        self.variants.push(MigrateVariant::#variant);
+                        Self {
+                            #(#fields_names),*
+                            ..self
+                        }
+                    }
+                }
+            })
+            .collect();
+
+        let dispatch_arms = methods
+            .iter()
+            .zip(migrate_variants.iter())
+            .map(|(method, variant)| {
+                let MsgMethod {
+                    function_name,
+                    fields,
+                    ..
+                } = method;
+                let fields_names = fields.iter().map(MsgField::name);
+                quote! {
+                    MigrateVariant :: #variant=> {
+                        contract.#function_name(ctx.branch().into(), #(#fields_names),*).map_err(|e| e.into())
+                    }
+                }
+            })
+            .collect::<Vec<_>>();
+
+        let result = match result {
+            ReturnType::Type(_, ty) => ty,
+            _ => unreachable!(),
+        };
+
+        quote! {
+            #[derive(#sylvia ::serde::Serialize, #sylvia ::serde::Deserialize, Clone, Debug, PartialEq, #sylvia ::schemars::JsonSchema)]
+            pub enum MigrateVariant {
+                #(#migrate_variants),*
+            }
+
+            #[allow(clippy::derive_partial_eq_without_eq)]
+            #[derive(#sylvia ::serde::Serialize, #sylvia ::serde::Deserialize, Clone, Debug, PartialEq, #sylvia ::schemars::JsonSchema)]
+            #[serde(rename_all="snake_case")]
+            pub struct MigrateMsg #generics #where_clause {
+                variants: Vec<MigrateVariant>,
+                #(pub Option<#fields>,)*
+            }
+
+            impl #generics MigrateMsg #generics #where_clause {
+                pub fn new(#(#parameters,)*) -> Self {
+                    Self { variants: vec![], #(#fields_names,)* }
+                }
+
+                #(#builders)*
+
+                pub fn dispatch #unused_generics(mut self, contract: &#contract_type, ctx: #ctx_type)
+                    -> #result #full_where
+                {
+                    let Self { variants, #(#fields_names,)* } = self;
+                    let mut ctx: #sylvia ::types::MigrateCtx = ctx.into();
+
+                    let mut resp = #sylvia ::cw_std::Response::new();
+                    let results: Vec< #result > = variants.iter().map(|v| match v {
+                        #(#dispatch_arms),*
+                    }).collect();
+                    for result in results {
+                        let result = result?;
+                        resp = resp.add_events(result.events);
+                        resp = resp.add_attributes(result.attributes);
+                    }
```

Comment:
This is the most uncertain part of the PR. Because we call multiple migrate methods returning multiple Results we should merge them somehow not to loose any info. Currently `data: Binary` is not concatenated. Also submsgs and messages are left.